design-proposals/kubebmc.md

```diff
@@ -0,0 +1,156 @@
+
+The following graph illustrates the entire KubeBMC workflow:
+
+```mermaid
+flowchart LR
+    controller["kubebmc-controller"]
+    cr["kubebmc CR"]
+    kbmc-pod["kbmc Pod"]
+    kbmc-svc["kbmc Service"]
+    controller-.->|watches|cr
+    cr-.->|owns|kbmc-svc
+    cr-.->|owns|kbmc-pod
+    client--->|IPMI|kbmc-svc
+    kbmc-svc-->kbmc-pod
+    kbmc-pod-->|HTTP|apiserver
+    apiserver-->|modifies|vm
+    vm-->|creates|vmi
+```
+
+It's also possible for VMs being created to opt out of such a functionality. This could be done by annotating the VirtualMachine object with a specific key-value pair.
+
+## API Examples
+
+The proposed KubeBMC CRD looks like the following:
+
+```yaml
+kind: KubeBMC
+metadata:
+  name: openshift-cluster-node-0
+spec:
+  # A Secret object reference for the IPMI/Redfish authentication.
+  # It should contain a pair of username and password.
+  authSecretRef:
+    namespace: default
+    name: bmc-cred
+  # The associated VM
+  virtualMachineRef:
+    namespace: default
+    name: openshift-cluster-node-0
+```
+
+The condition of the created resources (Pod and Service) will be reflected under the `.status` section:
+
+```yaml
+status:
+  serviceIP: 10.53.23.74
+  conditions:
+  - lastUpdateTime: "2024-03-01T08:48:23Z"
+    status: "True"
+    type: BMCEmulatorReady
+  - lastUpdateTime: "2024-03-01T08:48:23Z"
+    type: EndpointReady
+    status: "True"
+```
+
+## Scalability
+
+KubeBMC is designed to scale within Kubernetes environments, and the number of emulated BMC Pods and Services should increase as the number of VM increases. Hence, the overhead each Pod introduced to the entire system should also be considered.
+
+## Update/Rollback Compatibility
+
+The proposed feature does not introduce any breaking changes to KubeVirt itself, so the functionality of versions before this feature should not be impacted upon update/rollback.
+
+## Functional Testing Approach
+
+Utilizes Kubernetes testing frameworks to validate functionality across different deployment scenarios.
+
+Functional tests will simply:
+
+- Create a VM
+- Ensure the corresponding `kbmc` Pod and Serivce objects are created
+- Create a testing Pod with any container image that includes the `ipmitool`/`curl` binaries
+- Ensure the BMC emulator works as expected by sending IPMI/Redfish requests within the testing Pod to the BMC endpoint of the VM
+
+# Implementation Phases
+
+## Phase 1: Transfer the KubeBMC project to the KubeVirt organization
+
+The [KubeBMC project](https://github.com/starbops/kubebmc) already has a basic set of IPMI functionalities implemented:
+
+- Power on/off/reset
+- Power status retrieval
+- Boot device configuration (PXE or disk)
+
+However, some chores need to be done before transferring. For example, the API group of the KubeBMC CRD needs to be replaced with a suitable name conforming to KubeVirt's convention. Also, the code needs to be protected by robust unit tests and integration test plans.
```

Comment:
I would also consider renaming the CRDs. E.g. `VirtualMachineBMC`?

Reply:
Good idea! It makes more sense because many people get confused when they hear KubeBMC. "Is it an emulated BMC for the entire Kubernetes cluster?" something like that :P

Reply:
Exactly, I think this way it would better fit into the naming scheme of existing KubeVirt CRDs.